salt/utils/win_runas.py

```diff
@@ -38,428 +38,135 @@ def __virtual__():
-    # This only works as system, when salt is running as a service for example
-
-    # Check for a domain
+def split_username(username):
+    # TODO: Is there a windows api for this?
     domain = '.'
     if '@' in username:
         username, domain = username.split('@')
     if '\\' in username:
         domain, username = username.split('\\')
+    return username, domain
+
+
+def runas(cmdLine, username, password=None, cwd=None, elevated=True):
+
+    impersonation_token = salt.platform.win.impersonate_sid(
+        salt.platform.win.SYSTEM_SID,
+        session_id=0,
+        privs=['SeTcbPrivilege'],
+    )
+
+    username, domain = split_username(username)
+    sid, domain, sidType = win32security.LookupAccountName(domain, username)
+    if domain == 'NT AUTHORITY':
+        logonType = win32con.LOGON32_LOGON_SERVICE
+        user_token = win32security.LogonUser(
+            username,
+            domain,
+            '',
+            win32con.LOGON32_LOGON_SERVICE,
+            win32con.LOGON32_PROVIDER_DEFAULT,
+        )
+    elif password:
+        user_token = win32security.LogonUser(
+            username,
+            domain,
+            password,
+            win32con.LOGON32_LOGON_INTERACTIVE,
+            win32con.LOGON32_PROVIDER_DEFAULT,
+        )
+    else:
+        user_token = salt.platform.win.logon_msv1_s4u(username).Token
+
+    elevation_type = win32security.GetTokenInformation(
+        user_token, win32security.TokenElevationType
+    )
+    if elevation_type > 1:
+        user_token = win32security.GetTokenInformation(
+            user_token,
+            win32security.TokenLinkedToken
+        )
+        salt.platform.win.elevate_token(user_token)
+
+    handle_reg = win32profile.LoadUserProfile(user_token, {'UserName': username})
+    salt.platform.win.grant_winsta_and_desktop(user_token)
 
-    # Load User and Get Token
-    token = win32security.LogonUser(username,
-                                    domain,
-                                    password,
-                                    win32con.LOGON32_LOGON_INTERACTIVE,
-                                    win32con.LOGON32_PROVIDER_DEFAULT)
-
-    # Load the User Profile
-    handle_reg = win32profile.LoadUserProfile(token, {'UserName': username})
-
-    try:
-        # Get Unrestricted Token (UAC) if this is an Admin Account
-        elevated_token = win32security.GetTokenInformation(
-            token, win32security.TokenLinkedToken)
-
-        # Get list of privileges this token contains
-        privileges = win32security.GetTokenInformation(
-            elevated_token, win32security.TokenPrivileges)
-
-        # Create a set of all privileges to be enabled
-        enable_privs = set()
-        for luid, flags in privileges:
-            enable_privs.add((luid, win32con.SE_PRIVILEGE_ENABLED))
-
-        # Enable the privileges
-        win32security.AdjustTokenPrivileges(elevated_token, 0, enable_privs)
```

Comment:
You might still need this for some scripts to run (303 - 309). Even though it has the token, not all privileges are enabled. This just enables all of them.

Reply:
@twangboy Based on our Slack discussion. I agree and will put it back.

Reply:
Turns out this was moved to the `elevate_token` function.

Reply:
I guess I missed that.